crates/rrg/src/action/get_filesystem_timeline_tsk.rs

```diff
@@ -0,0 +1,582 @@
+        let mut fs = tsk::TskFs::open(&image)?;
+        let dir = fs.open_dir(&embedded_path)?;
+        let root_bytes = embedded_path.as_os_str().as_encoded_bytes();
+        fs.walk_dir(dir.addr(), |file, path| {
+            match tx.send(serialize_entry(root_bytes, path, file)) {
+                Ok(()) => tsk::WalkDirCallbackResult::Continue,
+                // rx disconnected, no need to keep searching.
+                Err(_) => tsk::WalkDirCallbackResult::Stop,
+            }
+        })
+    });
+
+    // `entry_count` keeps track of the number of entries that are included in
+    // each batch. Each time the `entries` iterator (defined below) yields an
+    // entry, we increase the count (through `Iterator::inspect`). We read the
+    // of `entry_count` when we are about to send a batch and reset the counter
+    // so that we start from 0 for the next batch.
+    //
+    // Note that `entry_count` has to be a cell because it is mutably borrowed
+    // by the `entries` iterator but we still want to be able to also modify it
+    // when we process batches.
+    let entry_count = &std::cell::RefCell::new(0);
+
+    let result_iter = rx.into_iter().inspect(|_| *entry_count.borrow_mut() += 1);
+
+    let encode_and_send_result = (move || -> crate::session::Result<()> {
+        for batch in crate::gzchunked::encode(result_iter) {
+            let batch = batch.map_err(crate::session::Error::action)?;
+
+            let blob = crate::blob::Blob::from(batch);
+            let blob_sha256 = sha2::Sha256::digest(blob.as_bytes()).into();
+
+            session.send(crate::Sink::Blob, blob)?;
+            session.reply(Item {
+                blob_sha256,
+                entry_count: *entry_count.borrow(),
+            })?;
+
+            *entry_count.borrow_mut() = 0;
+        }
+        Ok(())
+    })();
+
+    tsk_thread
+        .join()
+        .expect("tsk file walker thread panicked")
+        .map_err(Error::action)?;
+    // Check this after the thread joined.
+    encode_and_send_result?;
+
+    Ok(())
+}
+
+impl crate::request::Args for Args {
+    type Proto = rrg_proto::get_filesystem_timeline_tsk::Args;
+
+    fn from_proto(mut proto: Self::Proto) -> Result<Args, crate::request::ParseArgsError> {
+        use crate::request::ParseArgsError;
+
+        let raw_fs = PathBuf::try_from(proto.take_raw_fs())
+            .map_err(|error| ParseArgsError::invalid_field("raw_fs", error))?;
+        let root = PathBuf::try_from(proto.take_root())
+            .map_err(|error| ParseArgsError::invalid_field("root", error))?;
+
+        Ok(Args { raw_fs, root })
+    }
+}
+
+impl crate::response::Item for Item {
+    type Proto = rrg_proto::get_filesystem_timeline_tsk::Result;
+
+    fn into_proto(self) -> Self::Proto {
+        let mut proto = Self::Proto::default();
+        proto.set_blob_sha256(self.blob_sha256.into());
+        proto.set_entry_count(self.entry_count as u64);
+
+        proto
+    }
+}
+
+fn serialize_entry(
```

Comment:
I feel like this is not the best name for this function, when I saw it called above my first thought was "why do we serialize entries on the walking thread?" but it actually does not do any serialization, it just creates an entry.

Thus, I feel like `make_entry` or `create_entry` would fit much better.

Reply:
Changed to `make_entry`